Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/DownloadDiagnosisKeysTask.kt

```diff
@@ -0,0 +1,235 @@
+    }
+
+    private fun retrieveToken(rollbackItems: MutableList<RollbackItem>): String {
+        val googleAPITokenForRollback = LocalData.googleApiToken()
+        rollbackItems.add {
+            LocalData.googleApiToken(googleAPITokenForRollback)
+        }
+        return UUID.randomUUID().toString().also {
+            LocalData.googleApiToken(it)
+        }
+    }
+
+    private fun noKeysFetchedToday(): Boolean {
+        val currentDate = DateTime(timeStamper.nowUTC, DateTimeZone.UTC)
+        val lastFetch = DateTime(
+            LocalData.lastTimeDiagnosisKeysFromServerFetch(),
+            DateTimeZone.UTC
+        )
+        return (LocalData.lastTimeDiagnosisKeysFromServerFetch() == null ||
+            currentDate.withTimeAtStartOfDay() != lastFetch.withTimeAtStartOfDay()).also {
+            if (it) {
+                Timber.tag(TAG)
+                    .d("No keys fetched today yet (last=%s, now=%s)", lastFetch, currentDate)
+                BackgroundWorkHelper.sendDebugNotification(
+                    "Start Task",
+                    "No keys fetched today yet \n${DateTime.now()}\nUTC: $currentDate"
+                )
+            }
+        }
+    }
+
+    private fun rollback(rollbackItems: MutableList<RollbackItem>) {
+        try {
+            Timber.tag(TAG).d("Initiate Rollback")
+            for (rollbackItem: RollbackItem in rollbackItems) rollbackItem.invoke()
+        } catch (rollbackException: Exception) {
+            Timber.tag(TAG).e(rollbackException, "Rollback failed.")
+        }
+    }
+
+    private suspend fun getAvailableKeyFiles(requestedCountries: List<String>?): List<File> {
+        val availableKeyFiles =
+            keyFileDownloader.asyncFetchKeyFiles(if (environmentSetup.useEuropeKeyPackageFiles) {
+                listOf("EUR")
+            } else {
+                requestedCountries
+                    ?: appConfigProvider.getAppConfig().supportedCountries
+            }.map { LocationCode(it) })
+
+        if (availableKeyFiles.isEmpty()) {
+            Timber.tag(TAG).w("No keyfiles were available!")
+        }
+        return availableKeyFiles
+    }
+
+    private fun checkCancel() {
+        if (isCanceled) throw TaskCancellationException()
+    }
+
+    override suspend fun cancel() {
+        Timber.w("cancel() called.")
+        isCanceled = true
+    }
+
+    sealed class Progress : Task.Progress {
+        object ApiSubmissionStarted : Progress()
+        object ApiSubmissionFinished : Progress()
+
+        object KeyFilesDownloadStarted : Progress()
+        data class KeyFilesDownloadFinished(val keyCount: Int, val fileSize: Long) : Progress()
+
+        override val primaryMessage = this::class.java.simpleName.toLazyString()
+    }
+
+    class Arguments(
+        val requestedCountries: List<String>? = null,
+        val withConstraints: Boolean = false
+    ) : Task.Arguments
+
+    data class Config(
+        @Suppress("MagicNumber")
```

Comment:
Internally we have 3 votes to remove it and no vote to leave it, imho we could globally exclude this rule.
@ralfgehrer & @d4rken what do you think?

Reply:
If everyone is okay with that I'll create a PR (after this is merged), only for MagicNumber stuff..